File: flut/.flutter/lib/flutter/widgets/scroll_controller.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flutter/foundation/change_notifier.dart';

class FlutScrollController extends FlutChangeNotifier<ScrollController> {
  FlutScrollController.createFromData({
    required super.runtime,
    required super.data,
    required super.target,
  }) : super.createFromData();

  FlutScrollController.createFromObject({
    required super.runtime,
    required super.oid,
    required super.target,
  }) : super.createFromObject(type: 'ScrollController');

  static FlutScrollController flutCreate(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return FlutScrollController.createFromData(
      runtime: runtime,
      data: data,
      target: ScrollController(
        initialScrollOffset: runtime.unpackRequiredField<double>(
          data,
          'initialScrollOffset',
        ),
      ),
    );
  }

  @override
  dynamic getRawProperty(String property) {
    switch (property) {
      case 'offset':
        return flutTarget.offset;
      case 'hasClients':
        return flutTarget.hasClients;
      case 'position':
        return flutTarget.position;
    }
    return super.getRawProperty(property);
  }

  @override
  dynamic callMethod(
    String method,
    List<dynamic> args,
    Map<String, dynamic> kwargs,
  ) {
    switch (method) {
      case 'jumpTo':
        flutTarget.jumpTo((args[0] as num).toDouble());
        return null;
      case 'animateTo':
        flutTarget.animateTo(
          (args[0] as num).toDouble(),
          duration: runtime.decodeObject<Duration>(kwargs['duration'])!,
          curve: runtime.decodeObject<Curve>(kwargs['curve'])!,
        );
        return null;
    }
    return super.callMethod(method, args, kwargs);
  }
}

File: flut/flutter/widgets/scroll_controller.py
from flut.flutter.foundation.change_notifier import ChangeNotifier
from flut.dart import Duration
from flut.flutter.animation.curves import Curve, Curves

from .scroll_position import ScrollPosition


class ScrollController(ChangeNotifier):
    _flut_type = "ScrollController"

    def __init__(self, initialScrollOffset: float = 0.0):
        super().__init__()
        self._flut_create(props={"initialScrollOffset": initialScrollOffset})

    @property
    def offset(self) -> float:
        return self._flut_get("offset")

    @property
    def hasClients(self) -> bool:
        return self._flut_get("hasClients")

    @property
    def position(self) -> ScrollPosition | None:
        return self._flut_get("position")

    def jumpTo(self, value: float):
        self._flut_call("jumpTo", value)

    def dispose(self):
        self._flut_call("dispose")

    def animateTo(self, offset: float, *, duration: Duration, curve: Curve):
        self._flut_call(
            "animateTo",
            offset,
            duration=duration._flut_pack(),
            curve=curve._flut_pack(),
        )
